apps/bc/contracts/BC.sol

```diff
@@ -0,0 +1,353 @@
+        uint256 poolBalance;
+        uint256 totalSupply;
+        uint256 totalBuySpend;
+        uint256 totalBuyReturn;
+        uint256 totalSellSpend;
+        uint256 totalSellReturn;
+        mapping(address => uint256) buyers;
+        mapping(address => uint256) sellers;
+    }
+    mapping(address=>mapping(uint256 => Batch)) public batchesByCollateralToken;
+    mapping(address=>mapping(address => uint256[])) public addressToBlocksByCollateralToken;
+
+    event Mint(address indexed to, uint256 amount, address collateralToken);
+    event Burn(address indexed burner, uint256 amount, uint256 collateralToken);
+    event Buy(address indexed to, uint256 poolBalance, uint tokenSupply, uint256 amountTokens, uint256 totalCostCollateral, uint256 collateralToken);
+    event Sell(address indexed from, uint256 poolBalance, uint tokenSupply, uint256 amountTokens, uint256 returnedCollateral, uint256 collateralToken);
+
+    function initialize(
+        Pool _pool,
+        Vault _vault,
+        TokenManager _tokenManager,
+        address _bondingCurveToken,
+        address[] _collateralTokens,
+        uint256[] _virtualSupplies,
+        uint256[] _virtualBalances,
+        uint32[] _reserveRatios,
+        uint256 _batchBlocks) public onlyInit {
+        initialized();
+
+        require(isContract(_pool), ERROR_POOL_NOT_CONTRACT);
+        require(isContract(_vault), ERROR_VAULT_NOT_CONTRACT);
+        require(isContract(_tokenManager), ERROR_TM_NOT_CONTRACT);
+        require(isContract(_bondingCurveToken), ERROR_BCTOKEN_NOT_CONTRACT);
+        require(_batchBlocks >= 0, ERROR_INVALID_INIT_PARAMETER);
+        require(_collateralTokens.length <= MAX_COLLATERAL_TOKENS, ERROR_INVALID_INIT_PARAMETER);
+        uint256 len = _collateralTokens.length;
+        require(
+            len == _virtualSupplies.length && 
+            len == _virtualBalances.length &&
+            len == _reserveRatios.length, ERROR_INVALID_INIT_PARAMETER);
+        require(_collateralTokens.length >= 0, ERROR_INVALID_INIT_PARAMETER);
+        require(_virtualSupplies.length >= 0, ERROR_INVALID_INIT_PARAMETER);
+        require(_virtualBalances.length >= 0, ERROR_INVALID_INIT_PARAMETER);
+        require(_reserveRatios.length >= 0, ERROR_INVALID_INIT_PARAMETER);
+
+        pool = _pool;
+        vault = _vault;
+        bondingCurveToken = _bondingCurveToken;
+
+        // TODO: is it possible to use mappings as init parameters?
+        for (uint256 i = 0; i < _collateralTokens.length; i++) {
+            address collateralToken = _collateralTokens[i];
+            virtualSupplies[collateralToken] = _virtualSupplies[i];
+            virtualBalances[collateralToken] = _virtualBalances[i];
+            reserveRatios[collateralToken] = _reserveRatios[i];
+        }
+        collateralTokens = _collateralTokens;
+        batchBlocks = _batchBlocks;
+    }
+
+    /***** external function *****/
+    // function buy(uint256 _value) external payable;
```

Comment:
Why aren't these ones implemented yet? Is there a particular issue or do you wait more feedback on the remaining.